[FLINK-31157][release] Add release note for version 1.17 (#22146)
apache/flink: docs/content.zh/release-notes/flink-1.17.md

upgrade your serializer to `TypeSerializerSnapshot`, take a savepoint and restore from that 
savepoint in Flink >= 1.7.0.

### Python

#### Remove duplicated operators in the StreamGraph for Python DataStream API jobs
##### [FLINK-31272](https://issues.apache.org/jira/browse/FLINK-31272)
This duplicated operator issue has been addressed since 1.15.4, 1.16.2 and 1.17.0. For jobs which 
are not affected by this issue, there are no backward compatibility issues. However, for jobs which 
are affected, it may not be possible to restore from savepoints generated from versions 
1.15.0 ~ 1.15.3 and 1.16.0 ~ 1.16.1.

### Dependency upgrades

#### Upgrade FRocksDB to 6.20.3-ververica-2.0
##### [FLINK-30836](https://issues.apache.org/jira/browse/FLINK-30836)
Upgrade FRocksDB to 6.20.3-ververica-2.0.

#### Upgrade the minimal supported hadoop version to 2.10.2
##### [FLINK-29710](https://issues.apache.org/jira/browse/FLINK-29710)
The minimum Hadoop version supported by Apache Flink has been updated to version 2.10.2. For Hadoop 
3, the minimum Hadoop version that is now supported is version 3.2.3.

> #### Upgrade Calcite version to 1.27
> ##### [FLINK-20873](https://issues.apache.org/jira/browse/FLINK-20873)

Reviewer: Since finally in Flink 1.17.0 it was upgraded to 1.29.0, I think it would make sense to use the latest Calcite version in 1.17 branch
```suggestion
#### Upgrade Calcite version to 1.29.0
##### [FLINK-20873](https://issues.apache.org/jira/browse/FLINK-20873)
##### [FLINK-21239](https://issues.apache.org/jira/browse/FLINK-21239)
##### [FLINK-29932](https://issues.apache.org/jira/browse/FLINK-29932)
Calcite upgrade brings optimization changing logical plans for some queries involving `Sarg` and 
queries with `count` of non-distinct values.
```

---
title: "Release Notes - Flink 1.17"
---
<!--
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
-->

# Release notes - Flink 1.17

> These release notes discuss important aspects, such as configuration, behavior, or dependencies,

Reviewer: ```suggestion
These release notes discuss important aspects, such as configuration, behavior or dependencies,
```
not sure, however it looks like no need for comma here...
I try to follow this rule https://www.grammarly.com/blog/comma-before-or/

`MiniClusterWithClientResource` is a drop-in replacement.

#### Kinesis connector doesn't shade jackson dependency
##### [FLINK-14896](https://issues.apache.org/jira/browse/FLINK-14896)
Shade and relocate transitive Jackson dependencies of `flink-connector-kinesis`. If your Flink job 
was transitively relying on the these, you may need to include additional Jackson dependencies into 
your project.

### Runtime & Coordination

#### Speculative execution take input data amount into account when detecting slow tasks
##### [FLINK-30707](https://issues.apache.org/jira/browse/FLINK-30707)
The slow task detecting is improved for speculative execution. Previously, it only considers the 
execution time of tasks when deciding which tasks are slow. It now takes the input data volume of 
tasks into account. Tasks which has a longer execution time but consumes more data may not be 
considered as slow. This improvement helps to eliminate the negative impacts of data skew on slow 
task detecting.

#### Use adaptive batch scheduler as default scheduler for batch jobs
##### [FLINK-30682](https://issues.apache.org/jira/browse/FLINK-30682)

> Adaptive batch scheduler are now used for batch jobs by default. It will automatically decide the 

Reviewer 2: ```suggestion
Adaptive batch scheduler is now used for batch jobs by default. It will automatically decide the 
```

> parallelism of operators. The keys and values of related configuration items are improved for easy 

Reviewer 2: ```suggestion
parallelism of operators. The keys and values of related configuration items are improved for ease 
```

of use. More details can be found in the 
[document](https://nightlies.apache.org/flink/flink-docs-release-1.17/docs/deployment/elastic_scaling/#adaptive-batch-scheduler).

#### Simplify network memory configurations for TaskManager
##### [FLINK-30469](https://issues.apache.org/jira/browse/FLINK-30469)
The default value of `taskmanager.memory.network.max` has changed from `1g` to `Long#MAX_VALUE`, 

> to reduce the number of config options user needs to tune when trying to increase the network 

Reviewer 2: ```suggestion
to reduce the number of config options the user needs to tune when trying to increase the network 
```

memory size. This may affect the performance when this option is not explicitly configured, due to 
potential changes of network memory size, heap and managed memory size when the total memory 

> size is fixed. To go back to the previous behavior, user can explicitly configure this option to 

Reviewer 2: ```suggestion
size is fixed. To go back to the previous behavior, the user can explicitly configure this option to 
```

> the previous default value `1g`.

Reviewer 2: ```suggestion
the previous default value of `1g`.
```

A threshold is introduced for controlling the number of required buffers among all buffers needed 

> for reading data from upstream tasks. Reducing the number of required buffers helps reduces the 

Reviewer 2: ```suggestion
for reading data from upstream tasks. Reducing the number of required buffers helps reduce the 
```

unchanged for streaming workloads. This can be tuned via 
`taskmanager.network.memory.read-buffer.required-per-gate.max`. See the description of the config 
option for more details.

### Metric Reporters

#### Only support reporter factories for instantiation
##### [FLINK-24235](https://issues.apache.org/jira/browse/FLINK-24235)
Configuring reporters by their class is no longer supported. Reporter implementations must provide 
a MetricReporterFactory, and all configurations must be migrated to such a factory.

#### `UseLogicalIdentifier` makes datadog consider metric as custom
##### [FLINK-30383](https://issues.apache.org/jira/browse/FLINK-30383)
The Datadog reporter now adds a "flink." prefix to metric identifiers if "useLogicalIdentifier" is 
enabled. This is required for these metrics to be recognized as Flink metrics, not custom ones.

#### Use separate Prometheus CollectorRegistries
##### [FLINK-30020](https://issues.apache.org/jira/browse/FLINK-30020)
The PrometheusReporters now use a separate CollectorRegistry for each reporter instance instead of 
the singleton default registry. This generally shouldn't impact setups, but it may break code that 

> indirectly interacted with the reporter via the singleton instance (e.g., a test trying to assert 

Reviewer 2: ```suggestion
indirectly interacts with the reporter via the singleton instance (e.g., a test trying to assert 
```

what metrics are reported).

### Checkpoints

#### Drop TypeSerializerConfigSnapshot and savepoint support from Flink versions < 1.8.0
##### [FLINK-29807](https://issues.apache.org/jira/browse/FLINK-29807)

> Savepoints using `TypeSerializerConfigSnapshot` are no longer supported. That means, all savepoints 

Reviewer 2: ```suggestion
Savepoints using `TypeSerializerConfigSnapshot` are no longer supported. That means that all savepoints 
```

from Flink < 1.8.0 are no longer supported. Furthermore, savepoints from Flink < 1.17.0 created 
with custom serializer using deprecated since Flink 1.8.0 class `TypeSerializerConfigSnapshot` are 
also no longer supported.

If you are only using built-in serializers (Pojo, Kryo, Avro, 

> Tuple, ...), and your savepoint is from 1.8.0 <= Flink, you don't have to do anything.

Reviewer 2: ```suggestion
Tuple, ...), and your savepoint is from Flink >= 1.8.0, you don't have to do anything.
```

If you are only using built-in serializers (Pojo, Kryo, Avro, Tuple, ...), and your savepoint is 
from Flink < 1.8.0, please first upgrade your job to 1.8.x <= Flink <= 1.16.x, before upgrading in 

> the second step to Flink >= 1.17.x.

Reviewer 2: ```suggestion
a second step to Flink >= 1.17.x.
```

If previously you were using a custom serialiser that 

> depends on `TypeSerializerConfigSnapshot`, please first while still using 1.8.x <= Flink <= 1.16.x 
> upgrade your serializer to `TypeSerializerSnapshot`, take a savepoint and restore from that 

Reviewer 2: ```suggestion
depends on `TypeSerializerConfigSnapshot`, please upgrade your serializer to `TypeSerializerSnapshot` while still using 1.8.x <= Flink <= 1.16.x, take a savepoint and restore from that 
```

savepoint in Flink >= 1.7.0.

### Python

#### Remove duplicated operators in the StreamGraph for Python DataStream API jobs
##### [FLINK-31272](https://issues.apache.org/jira/browse/FLINK-31272)
This duplicated operator issue has been addressed since 1.15.4, 1.16.2 and 1.17.0. For jobs which 
are not affected by this issue, there are no backward compatibility issues. However, for jobs which 
are affected, it may not be possible to restore from savepoints generated from versions 
1.15.0 ~ 1.15.3 and 1.16.0 ~ 1.16.1.

> 

Reviewer 3: ```suggestion
#### Adds support of Python 3.10 and removes support of Python 3.6
##### [FLINK-29421](https://issues.apache.org/jira/browse/FLINK-29421)
PyFlink 1.17 will support Python 3.10 and remove the support of Python 3.6
```

### Dependency upgrades

#### Upgrade FRocksDB to 6.20.3-ververica-2.0
##### [FLINK-30836](https://issues.apache.org/jira/browse/FLINK-30836)
Upgrade FRocksDB to 6.20.3-ververica-2.0.

#### Upgrade the minimal supported hadoop version to 2.10.2
##### [FLINK-29710](https://issues.apache.org/jira/browse/FLINK-29710)
The minimum Hadoop version supported by Apache Flink has been updated to version 2.10.2. For Hadoop 
3, the minimum Hadoop version that is now supported is version 3.2.3.

> 

Reviewer 3: ```suggestion
#### Update dependency version for PyFlink
##### [FLINK-29421](https://issues.apache.org/jira/browse/FLINK-29421)
For support of Python3.10, PyFlink updates a series dependencies version: apache-beam==2.43.0 pemja==0.3.0 cloudpickle==2.2.0
#####
```

> #### Upgrade Calcite version to 1.29.0

Reviewer 4: It is mentioned in the "Calcite Upgrade" section of Flink 1.17 release announcement that "The downside of this upgrade is that the execution plan changes for certain SQL queries. Savepoints are not backwards compatible in these cases".

Would this prevent users from upgrading from Flink 1.16 to Flink 1.17 without losing job state? Is there specific actions that users should take to get around this issue? Maybe we should mention this in the release note.

Reviewer 5: +1

Reviewer 5: > It is mentioned in the "Calcite Upgrade" section of Flink 1.17 release announcement that "The downside of this upgrade is that the execution plan changes for certain SQL queries. Savepoints are not backwards compatible in these cases".
> 
> Would this prevent users from upgrading from Flink 1.16 to Flink 1.17 without losing job state? Is there specific actions that users should take to get around this issue? Maybe we should mention this in the release note.

+1, @PatrickRen You can refer previous release which also upgrade the  calcite version

Author: Thanks for reviewing this @lindong28 @leonardBang !

> Would this prevent users from upgrading from Flink 1.16 to Flink 1.17 without losing job state? Is there specific actions that users should take to get around this issue? Maybe we should mention this in the release note.

@lindong28 TBH I'm not quite familiar with this part. Maybe @snuyanzin could provide some inputs on the migration path. But I could add some warns in this section that highlights the incompatibility.

> You can refer previous releases which also upgrade the calcite version

@leonardBang Unfortunately I can't find previous release notes mentioning upgrading Calcite.

Reviewer: As far as I know there is currently no full support of backward compatibility for Table SQL/API while major version upgrade and this is a topic for FLIP-190 [1].
A good side here is that there are changes only for logical plans. Neither physical plans nor ser/de for json plans have been impacted. That means that ser/de from json plans while upgrade should work. 
The only issue is that i fail to find a proper doc for that... Only a dedicated jira issue[2]

[1] https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=191336489
[2] https://issues.apache.org/jira/browse/FLINK-26250

Reviewer 4: One of the most important purposes of release notes is to explain to users how to upgrade existing jobs from Flink 1.16 (and ideally earlier versions as well) to Flink 1.17. Without good upgrade instructions, the upgrade experience may be bad and many users may be scared away from using Flink 1.17.

In this specific case, I hope that users can find the following information from the release notes:

- Will I be able to restart a job successfully with Flink 1.17 using the savepoint taken when the job was executed with Flink 1.16?
- If not, do I have to discard state and start the job from a very early point to re-process data? What impact should I be prepared for when upgrading to Flink 1.17?

Reviewer: Sorry for the confusion...
Based on the current doc [1] 
```
Currently, the framework cannot guarantee that state can be mapped from a savepoint to a new table operator topology.

In other words: Savepoints are only supported if both the query and the Flink version remain constant.
```
Also we tried to do such test for simple SQL jobs to restore in 1.17 from the state made in 1.16 and failed [2], same issue between 1.15 and 1.16.
Since this is the current state of the upgrade between major version and not Calcite upgrade only related feature, 
does it make sense to remove this sentence `"The downside of this upgrade is that the execution plan changes for certain SQL queries. Savepoints are not backwards compatible in these cases"`? Because anyway it's mentioned explicitly in doc [1]

[1] https://nightlies.apache.org/flink/flink-docs-master/docs/dev/table/concepts/overview/#stateful-upgrades-and-evolution
[2] https://issues.apache.org/jira/browse/FLINK-31066

Author: Thanks @snuyanzin for the context. As the [document](https://nightlies.apache.org/flink/flink-docs-master/docs/dev/table/concepts/overview/#stateful-upgrades-and-evolution) I think it's not easy to give such a migration path described by @lindong28, because we couldn't tell exactly which kind of query or plan is changed by the Calcite upgrade. This incompatibility issue should have already existed for many versions as optimizing rules keep evolving. 

+1 for removing the statement from the announcement.

Reviewer 4: Given that Flink does not guarantee savepoint compatibility between major/minor releases, I think removing that sentence from the release announcement is reasonable.

##### [FLINK-20873](https://issues.apache.org/jira/browse/FLINK-20873)
##### [FLINK-21239](https://issues.apache.org/jira/browse/FLINK-21239)
##### [FLINK-29932](https://issues.apache.org/jira/browse/FLINK-29932)

> Calcite upgrade brings optimization changing logical plans for some queries involving `Sarg` and 

Reviewer 2: ```suggestion
Calcite upgrade brings optimizations that change logical plans for some queries involving `Sarg` and 
```
